File: src/Billing/PollService.php
<?php
declare(strict_types=1);
namespace App\Billing;
use App\Infrastructure\{Database,Outbox};
final class PollService
{
    public function __construct(private Database $db, private Outbox $outbox, private Wallet $wallet) {}
    /** Create a poll with questions. */
    public function create(array $input, string $actor): array
    {
        $title = trim((string)($input['title'] ?? ''));
        if ($title === '' || mb_strlen($title) > 255) throw new BillingError('Заголовок: 1–255 символов.');
        $reward = (int)($input['reward_amount_kopeks'] ?? 0);
        if ($reward<0 || $reward>100000000) throw new BillingError('Награда: от 0 до 1 000 000 ₽.');
        $questions = $input['questions'] ?? [];
        if (!is_array($questions) || count($questions) < 1 || count($questions) > 10) throw new BillingError('Добавьте 1–10 вопросов.');
        $id = Database::id();
        $now = time();
        $this->db->transaction(function () use ($id, $title, $reward, $questions, $actor, $now, $input) {
            $this->db->execute('INSERT INTO polls(id,title,description,reward_enabled,reward_amount_kopeks,created_by,created_at) VALUES(?,?,?,?,?,?,?)', [$id, $title, $input['description'] ?? null, (int)($reward > 0), $reward, $actor, $now]);
            foreach ($questions as $i => $q) {
                $options = $q['options'] ?? [];
                $question=trim((string)($q['text'] ?? ''));
                if ($question==='' || mb_strlen($question)>500 || !is_array($options) || count($options)<2 || count($options)>20) throw new BillingError('Проверьте вопрос и его варианты.');
                foreach ($options as $option) if (!is_string($option)) throw new BillingError('Варианты ответа должны быть текстом.');
                $options=array_map(static fn(string $option)=>trim($option),array_values($options));
                if (count(array_filter($options,static fn($option)=>$option!=='' && mb_strlen($option)<=500))!==count($options) || count(array_unique($options))!==count($options)) throw new BillingError('Варианты ответа должны быть уникальными и непустыми.');
                $this->db->execute('INSERT INTO poll_questions(id,poll_id,text,options,"order") VALUES(?,?,?,?,?)', [Database::id(), $id, $question, json_encode($options, JSON_THROW_ON_ERROR), $i]);
            }
            $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'poll.created', $id, $now]);
        });
        return $this->db->one('SELECT * FROM polls WHERE id=?', [$id]);
    }
    public function list(): array
    {
        return $this->db->all('SELECT * FROM polls ORDER BY created_at DESC');
    }
    public function get(string $id): ?array
    {
        $poll = $this->db->one('SELECT * FROM polls WHERE id=?', [$id]);
        if (!$poll) return null;
        $poll['questions'] = $this->db->all('SELECT * FROM poll_questions WHERE poll_id=? ORDER BY "order"', [$id]);
        return $poll;
    }
    /** Submit answers. Returns ['reward'=>int]. */
    public function submit(string $userId, string $pollId, array $answers): array
    {
        return $this->db->transaction(function () use ($userId, $pollId, $answers) {
            $poll = $this->db->one('SELECT * FROM polls WHERE id=?' . $this->db->lock(), [$pollId]);
            if (!$poll) throw new BillingError('Опрос не найден.');
            $user=$this->db->one('SELECT disabled FROM users WHERE id=?',[$userId]);
            if (!$user || (int)$user['disabled']===1) throw new BillingError('Аккаунт недоступен.');
            if ($this->db->one('SELECT id FROM poll_responses WHERE poll_id=? AND user_id=?', [$pollId, $userId])) throw new BillingError('Вы уже участвовали в этом опросе.');
            $questions = $this->db->all('SELECT * FROM poll_questions WHERE poll_id=? ORDER BY "order"', [$pollId]);
            $normalized = [];
            foreach ($questions as $i => $q) {
                $answer = $answers[$i] ?? null;
                $options=json_decode($q['options'],true,512,JSON_THROW_ON_ERROR);
                if (!is_string($answer) || !in_array($answer,$options,true)) throw new BillingError('Выберите один из вариантов ответа.');
                $normalized[] = $answer;
            }
            $reward = 0;
            if ((int)$poll['reward_enabled'] === 1 && (int)$poll['reward_amount_kopeks'] > 0) {
                $this->wallet->credit($userId, (int)$poll['reward_amount_kopeks'], 'referral_reward', 'Награда за опрос: '.$poll['title']);
                $reward = (int)$poll['reward_amount_kopeks'];
            }
            $this->db->execute('INSERT INTO poll_responses(id,poll_id,user_id,answers,reward_paid,created_at) VALUES(?,?,?,?,?,?)', [Database::id(), $pollId, $userId, json_encode($normalized, JSON_THROW_ON_ERROR), (int)($reward > 0), time()]);
            return ['reward' => $reward];
        });
    }
}

File: rails_backend/app/services/billing/poll_service.rb
module Billing
  class PollService
    def create(input:, actor:)
      attrs = input.to_h.symbolize_keys
      title = attrs[:title].to_s.strip
      reward = attrs[:reward_amount_kopeks].to_i
      questions = attrs[:questions]
      raise BillingError, "Заголовок: 1–255 символов." unless title.length.between?(1, 255)
      raise BillingError, "Награда: от 0 до 1 000 000 ₽." unless reward.between?(0, 100_000_000)
      raise BillingError, "Добавьте 1–10 вопросов." unless questions.is_a?(Array) && questions.length.between?(1, 10)

      Poll.transaction do
        poll = Poll.create!(id: Infrastructure::IdGenerator.call, title: title,
          description: attrs[:description], reward_enabled: reward.positive? ? 1 : 0,
          reward_amount_kopeks: reward, created_by: actor, created_at: Time.now.to_i)
        questions.each_with_index do |raw, index|
          q = raw.to_h.symbolize_keys
          text = q[:text].to_s.strip
          options = q[:options]
          raise BillingError, "Проверьте вопрос и его варианты." unless text.length.between?(1, 500) && options.is_a?(Array) && options.length.between?(2, 20)
          options = options.map { |option| raise(BillingError, "Варианты ответа должны быть текстом.") unless option.is_a?(String); option.strip }
          raise BillingError, "Варианты ответа должны быть уникальными и непустыми." if options.any? { |option| option.empty? || option.length > 500 } || options.uniq.length != options.length
          PollQuestion.create!(id: Infrastructure::IdGenerator.call, poll_id: poll.id, text: text,
            options: options.to_json, order: index)
        end
        AuditLog.create!(id: Infrastructure::IdGenerator.call, actor: actor, action: "poll.created", subject: poll.id, created_at: Time.now.to_i)
        poll
      end
    end

    def list
      Poll.includes(:questions).order(created_at: :desc).limit(500)
    end

    def submit(user_id:, poll_id:, answers:)
      Poll.transaction do
        poll = Poll.lock.find_by(id: poll_id)
        raise BillingError, "Опрос не найден." unless poll
        user = User.find_by(id: user_id)
        raise BillingError, "Аккаунт недоступен." unless user && user.disabled.to_i.zero?
        raise BillingError, "Вы уже участвовали в этом опросе." if PollResponse.exists?(poll_id: poll.id, user_id: user_id)
        questions = poll.questions.order(:order).to_a
        normalized = questions.each_with_index.map do |question, index|
          answer = answers.is_a?(Array) ? answers[index] : nil
          options = JSON.parse(question.options)
          raise BillingError, "Выберите один из вариантов ответа." unless answer.is_a?(String) && options.include?(answer)
          answer
        end
        reward = poll.reward_enabled.to_i == 1 ? poll.reward_amount_kopeks.to_i : 0
        WalletService.new.credit(user_id, reward, "referral_reward", "Награда за опрос: #{poll.title}") if reward.positive?
        PollResponse.create!(id: Infrastructure::IdGenerator.call, poll_id: poll.id, user_id: user_id,
          answers: normalized.to_json, reward_paid: reward.positive? ? 1 : 0, created_at: Time.now.to_i)
        { reward: reward }
      end
    end
  end
end
